packages/console/src/pages/UserDetails/index.tsx

```diff
@@ -1,32 +1,101 @@
 import useSWR from 'swr';
 
 import BackLink from '@/components/BackLink';
+import Button from '@/components/Button';
 import Card from '@/components/Card';
+import CodeEditor from '@/components/CodeEditor';
 import CopyToClipboard from '@/components/CopyToClipboard';
+import FormField from '@/components/FormField';
 import ImagePlaceholder from '@/components/ImagePlaceholder';
-import { RequestError } from '@/hooks/use-api';
+import TabNav, { TabNavLink } from '@/components/TabNav';
+import TextInput from '@/components/TextInput';
+import useApi, { RequestError } from '@/hooks/use-api';
+import { safeParseJson } from '@/utilities/json';
 
 import CreateSuccess from './components/CreateSuccess';
 import * as styles from './index.module.scss';
 
+type FormData = {
+  primaryEmail: string;
+  primaryPhone: string;
+  username: string;
+  name: string;
+  avatar: string;
+  roles: string;
+  customData: string;
+};
+
 const UserDetails = () => {
   const { id } = useParams();
   const { t } = useTranslation(undefined, { keyPrefix: 'admin_console' });
 
-  const { data, error } = useSWR<User, RequestError>(id && `/api/users/${id}`);
+  const { data, error, mutate } = useSWR<User, RequestError>(id && `/api/users/${id}`);
   const isLoading = !data && !error;
 
+  const { handleSubmit, register, control, reset } = useForm<FormData>();
+  const [submitting, setSubmitting] = useState(false);
+  const {
+    field: { onChange, value },
+  } = useController({ name: 'customData', control, rules: { required: true } });
+  const api = useApi();
+
+  useEffect(() => {
+    if (!data) {
+      return;
+    }
+    reset({
+      primaryEmail: data.primaryEmail ?? '',
+      primaryPhone: data.primaryPhone ?? '',
+      username: data.username ?? '',
+      name: data.name ?? '',
+      avatar: data.avatar ?? '',
+      roles: data.roleNames.join(','),
+      customData: JSON.stringify(data.customData, null, 2),
+    });
+  }, [data, reset]);
+
+  const onSubmit = handleSubmit(async (formData) => {
+    if (!data || submitting) {
+      return;
+    }
+
+    const customData = safeParseJson(formData.customData);
+
+    if (!customData) {
+      toast.error(t('user_details.custom_data_invalid'));
+
+      return;
+    }
+
+    const payload: Partial<User> = {
+      name: formData.name || undefined,
+      avatar: formData.avatar || undefined,
+      customData,
+    };
+    setSubmitting(true);
+
+    try {
+      const updatedUser = await api.patch(`/api/users/${data.id}`, { json: payload }).json<User>();
+      void mutate(updatedUser);
+      toast.success(t('user_details.saved'));
+    } finally {
+      setSubmitting(false);
+    }
+  });
+
   return (
     <div className={styles.container}>
       <BackLink to="/users">{t('user_details.back_to_users')}</BackLink>
 
       {isLoading && <div>loading</div>}
       {error && <div>{`error occurred: ${error.metadata.code}`}</div>}
-      {data && (
+      {id && data && (
```

Comment:
Would be better if the form is visible even if `id` or `data` is empty

Reply:
Maybe we can optimized it after "Loading" design is ready?

Reply:
ok